dbms/src/IO/WriteBuffer.h

```diff
@@ -35,7 +35,7 @@ class WriteBuffer : public BufferBase
       */
     inline void next()
     {
-        if (!offset())
+        if (!offset() && available())
```

Comment:
@abyss7 can you elaborate this hunk?

Reply:
I suggest that I wanted to prevent miscalling `next()` if there is still available space in buffer - it's permissive, so no penalties for logical error in the code. But maybe there should be `assert(!available())`. Does this break anything?

Reply:
>Does this break anything?

I have't any example, but is this strictly correct?
AFAICS right now destructor tries to send the data, so after this change it will send only if the buffer is full, sounds icky

I came to this while working on #19451 
And quite frankly to me sending some data from destructor looks not safe anyway, and the goal is to replace destructor final flush with assert that will ensure that there is no bytes pending in buffer.

Reply:
```
if (!offset() && available())
    return;
```
This is strange.

This code means: if there are no data to write - ignore the `next` call.
```
if (!offset())
    return;
```

`available()` means - if the buffer is not filled up completely.
Adding this condition with "and" is bogus - because if the buffer size is greater than zero, and `!offset()` - at least one byte is available.

Even if you meant this:
```
if (!offset() || available())
    return;
```
it also does not make sense, because flushing half-filled buffer is completely legit and is used extensively.

Reply:
Okay, let's try to revert this hunk - #19886, and see, maybe CI will find something interesting.